File: leetcode/93-restore-ip-address/main.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

func restoreIpAddresses(s string) []string {
	ips := make([]string, 0)

	parts := make([]string, 4)
	for i := 1; i < min(4, len(s)); i++ {
		parts[0] = s[:i]
		if !isValid(parts[0]) {
			continue
		}
		for j := 1; j < min(4, len(s)-i); j++ {
			parts[1] = s[i : i+j]
			if !isValid(parts[1]) {
				continue
			}
			for k := 1; k < min(4, len(s)-i-j); k++ {
				parts[2] = s[i+j : i+j+k]
				if !isValid(parts[2]) {
					continue
				}
				parts[3] = s[i+j+k:]
				if !isValid(parts[3]) {
					continue
				}
				ips = append(ips, strings.Join(parts, "."))
			}
		}
	}

	return ips
}

func isValid(part string) bool {
	if len(part) > 1 && part[0] == '0' {
		return false
	}
	raw, _ := strconv.Atoi(part)
	if raw > 255 {
		return false
	}
	return true
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func main() {
	s := "010010"
	fmt.Println(restoreIpAddresses(s))
}

File: leetcode/93-restore-ip-address/main.py
from typing import List


class Solution:
    def is_valid(self, s: str) -> bool:
        if len(s) > 1 and s[0] == '0' or int(s) > 255:
            return False
        return True

    def restoreIpAddresses(self, s: str) -> List[str]:
        ips = []
        parts = [""] * 4
        for i in range(1, min(4, len(s))):
            parts[0] = s[:i]
            if not self.is_valid(parts[0]):
                continue
            for j in range(1, min(4, len(s) - i)):
                parts[1] = s[i:i+j]
                if not self.is_valid(parts[1]):
                    continue
                for k in range(1, min(4, len(s) - i - j)):
                    parts[2] = s[i + j:i + j + k]
                    if not self.is_valid(parts[2]):
                        continue
                    parts[3] = s[i + j + k:]
                    if not self.is_valid(parts[3]):
                        continue
                    ips.append(".".join(parts))
        return ips


solution = Solution()
print(solution.restoreIpAddresses("25525511135"))
